### PR TITLE
Allow a reason and extras when creating result from Nack exception

### PR DESCRIPTION
**Pull request by kcox-globality**
Allow an optional `reason` and `extras` when creating a result from a Nack exception. This will allow us to include additional information in logs when Nacks are raised by handlers -- which become RETRIED messages.

Here's an example using `reason` (see the `log_reason` field):
Screen Shot 2022-07-25 at 5 16 40 PM
^^ from https://github.com/globality-corp/euclides/pull/463

Closes [GLOB-66164](https://globality.atlassian.net/browse/GLOB-66164)

### CODE REVIEW

**Review by jroundsglobality — APPROVED** [review]

**Review by mchen-globality — APPROVED** [review]

File: microcosm_pubsub/errors.py [hunk 1]
```diff
@@ -20,8 +20,10 @@ class Nack(Exception):
     and the SQS dead letter queue (DLQ) configuration, if any.
 
     """
-    def __init__(self, visibility_timeout_seconds):
+    def __init__(self, visibility_timeout_seconds, reason="", extra=None):
```
Comment (sethisernhagen) [inline microcosm_pubsub/errors.py:23]: Can reason be None instead of empty string? Then function signature would match `TTLExpired()`.

Comment (sethisernhagen) [reply microcosm_pubsub/errors.py:23]: Interesting that `reason` is required for `SkipMessage`. Nack should probably be like that someday.

File: microcosm_pubsub/errors.py [hunk 2]
```diff
@@ -20,8 +20,10 @@ class Nack(Exception):
...
+        super().__init__(reason)
         self.visibility_timeout_seconds = visibility_timeout_seconds
+        self.extra = extra or dict()
```
Comment (sethisernhagen) [inline microcosm_pubsub/errors.py:26]: Should we update `__repr__` to match contructor?

File: microcosm_pubsub/tests/test_result.py [hunk 3]
```diff
@@ -153,6 +153,32 @@ def handler(message):
...
+                ),
+            ),
+        )
+        # nack with custom retry visibility timeout
+        self.graph.sqs_consumer.sqs_client.change_message_visibility.assert_called_with(
+            QueueUrl="queue",
+            ReceiptHandle=RECEIPT_HANDLE,
+            VisibilityTimeout=3,
+        )
+
```
Comment (sethisernhagen) [inline microcosm_pubsub/tests/test_result.py:181]: Nice!

**Review by sethisernhagen — APPROVED** [review]
This should have been this way for a long time.

File: setup.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
     install_requires=[
         "boto3>=1.5.8",
         "dataclasses;python_version<'3.7'",
+        "flake8<=4.0.1"
```
Comment (tonybajan-globality) [inline setup.py:23]: let's not pin this here. It's not a runtime dependency and will prevent all downstream dependants from upgrading at their own pace.  Once this branch is updated from master this should appear in the "lint" extra instead.

And I guess we hope the maintainers of [flake8-logging-format](https://github.com/globality-corp/flake8-logging-format) fix their builds 😳?

**Pull request merged**